File: DSA/Cpp/src/dsa/lib/math/ds/parentheses.cpp
#include <stack>
#include <map>
#include "parentheses.h"

namespace dsa::lib::math::ds {

bool Parentheses::isValid(std::string s) {
    std::stack<char> temp;
    std::map<char, char> m = {{']', '['}, {')', '('}, {'}', '{'}};
    for (int i = 0; i < s.size(); i++) {
        if (s[i] == '[' || s[i] == '(' || s[i] == '{')
            temp.push(s[i]);
        else if (s[i] == ']' || s[i] == ')' || s[i] == '}') {
            if (temp.empty() || temp.top() != m[s[i]]) return false;
            else temp.pop();
        }
    }
    return temp.empty();
}

}

File: DSA/Java/src/dsa/lib/math/ds/Parentheses.java
package dsa.lib.math.ds;

import java.util.HashMap;
import java.util.Map;
import java.util.Stack;

public class Parentheses {
    //    LeetCode20
    public boolean isValid(String s) {
        return isValid21(s);
    }

    //    Stack and Hash Map
    public static boolean isValid1(String s) {
//        Initialize the parentheses map
        Map<Character, Character> parenMap = new HashMap<>();
        parenMap.put(')', '(');
        parenMap.put('}', '{');
        parenMap.put(']', '[');

        Stack<Character> stack = new Stack<>();
        for (char c : s.toCharArray()) {
            // It is an opening bracket, push to the stack.
            if (!parenMap.containsKey(c)) stack.push(c);
            else if (stack.isEmpty() || stack.pop() != parenMap.get(c)) return false;
        }
        // If the stack is empty, it is valid.
        return stack.isEmpty();
    }

    //    Stack
    private static boolean isValid2(String s) {
        Stack<Character> stack = new Stack<>();
        for (char c : s.toCharArray()) {
//            Opening bracket
            if (c == '(') stack.push(')');
            else if (c == '{') stack.push('}');
            else if (c == '[') stack.push(']');
//            Closing bracket
            else if (stack.isEmpty() || stack.pop() != c) return false;
        }
        return stack.isEmpty();
    }

    private static boolean isValid21(String s) {
        Stack<Character> stack = new Stack<>();
        for (char c : s.toCharArray()) {
            switch (c) {
                case '(':
                    stack.push(')');
                    break;
                case '[':
                    stack.push(']');
                    break;
                case '{':
                    stack.push('}');
                    break;
                default:
                    if (stack.isEmpty() || stack.pop() != c) return false;
            }
        }
        return stack.isEmpty();
    }
}
